src/System.Management.Automation/engine/CommandProcessorBase.cs

```diff
@@ -644,13 +644,86 @@ internal void DoComplete()
                 _context.ShellFunctionErrorOutputPipe = oldErrorOutputPipe;
                 _context.CurrentCommandProcessor = oldCurrentCommandProcessor;
 
+                // Restore the previous scope...
+                if (_previousScope != null)
+                {
+                    // Restore the scope but use the same session state instance we
+                    // got it from because the command may have changed the execution context
+                    // session state...
+                    CommandSessionState.CurrentScope = _previousScope;
+                }
+
+                // Restore the previous session state
+                if (_previousCommandSessionState != null)
+                {
+                    Context.EngineSessionState = _previousCommandSessionState;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Executes the appropriate disposal methods for script commands.
+        /// <see cref="DlrScriptCommandProcessor"/> is itself a command processor, so interpreted functions cannot be
+        /// reliably disposed without also disposing the entire command processor. Instead, it implements its own
+        /// <see cref="DlrScriptCommandProcessor.InvokeCleanupBlock"/> method to perform the same function as its
+        /// compiled counterpart <see cref="PSScriptCmdlet.Dispose"/>, which is a proper IDisposable implementation.
+        /// </summary>
+        internal virtual void CleanupScriptCommand()
+        {
+            var scriptCmdlet = Command as PSScriptCmdlet;
+            var scriptProcessor = this as DlrScriptCommandProcessor;
+
+            if (scriptCmdlet == null && scriptProcessor == null)
+            {
+                return;
+            }
+
+            bool isStopping = ExceptionHandlingOps.SuspendStoppingPipeline(Context);
+            Pipe oldErrorOutputPipe = Context.ShellFunctionErrorOutputPipe;
+            CommandProcessorBase oldCurrentCommandProcessor = Context.CurrentCommandProcessor;
+
+            try
+            {
+                if (this.RedirectShellErrorOutputPipe || Context.ShellFunctionErrorOutputPipe != null)
+                {
+                    Context.ShellFunctionErrorOutputPipe = CommandRuntime.ErrorOutputPipe;
+                }
+
+                Context.CurrentCommandProcessor = this;
+                SetCurrentScopeToExecutionScope();
+
+                using (CommandRuntime.AllowThisCommandToWrite(true))
```

Comment:
I'll add the param name. It's named `permittedToWriteToPipeline`, but if my memory is correct if I set this to `false`, non-output streams can't be written to? Something like that, apologies, it's been a while since I was tinkering with it. 🙂